### DISCUSSION TITLE
php (not fpm) with BREF_LOOP_MAX=1  is 2x slower than php-fpm

### DISCUSSION THREAD

**Discussion by allan-simon** [brefphp/bref#1820] [Support]
**Description:**

Due to the issue #1369 , we've switched our "web" lambda (with symfony) from php-fpm layer to php (the same we use for symfony commands / messenger) 

We've noticed a significant increase in response time as you can see on our monitoring, (you can see the before/after switching very clearly on the response percentile graph)  so it does not seems related to a specific kind of request 

![image](https://github.com/brefphp/bref/assets/213167/e675f141-29c4-41b2-b095-08f8657630c7)

we're currently using  with BREF_LOOP_MAX=1 due to some symfony service being too stateful and not reset-ing correctly between two requests. 

However my understanding would be that bref without php-fpm should be at least as fast as with php-fpm (and faster when reusing the same kernel ) so I can't wrap  my head on why it's two times slower.

Any idea on where I can dig to find the core issue ?

**Comment by mnapoli** [comment]
Hi, this is expected because the function runtime (without the loop) restarts a new PHP process on every request. That's not really efficient for synchronous use-cases: https://bref.sh/docs/environment/performances#bref-for-event-driven-functions

PHP-FPM does not restart a process, it's highly optimized and wipes the process' memory basically.

For your scenario why not keep using FPM?

**Reply by mnapoli** [reply]
Oh sorry just noticed the intro. Might be worth trying to fix the original issue then, maybe by playing with the php-fpm.conf settings.

**Comment by allan-simon** [comment]
No worry, thanks for the explanation it makes sense, if  I understand correctly it's basically because when the lambda is hot php-fpm has already a process ready while with the loop set to 1 we wait each for the process to start and for the process to stop "cleanly" ?

THen my mental model was wrong , I thought that the "php" layer was one with less layer of abstraction , like directly   "php running bref , directly booting symfony in the same process" , while php-fpm was kind of "php running bref,  starting a php-fpm process that itself pop a php process with symfony" .  And in that mental model it made no sense that php was slower with less inter-process communication. But now I see clearer 🙂

I will try to dig more into the php-fpm issue then , this bug has been waiting for far too long :P

________________________________
